### PR TITLE
Make json assertions stricter

### PR DESCRIPTION
**Pull request by kalenp**
While writing a regression test for our API output, we discovered that the assertJson* methods allow some strings to pass which are not strictly equal.  These cases are covered particularly in the `string type not equals number` and `string type not equals boolean` test cases which have been added.  The additional tests help to validate behavior for any Json comparison method and in particular the proposed change.

### CODE REVIEW

File: src/Framework/Assert.php [hunk 1]
```diff
@@ -2242,10 +2242,12 @@ public static function assertJsonStringEqualsJsonString($expectedJson, $actualJs
         self::assertJson($expectedJson, $message);
         self::assertJson($actualJson, $message);
 
-        $expected = json_decode($expectedJson);
-        $actual   = json_decode($actualJson);
+        // call constraint
```
Comment (sebastianbergmann) [inline src/Framework/Assert.php]: We don't use inline comments like this.

File: src/Framework/Assert.php [hunk 2]
```diff
@@ -2260,10 +2262,12 @@ public static function assertJsonStringNotEqualsJsonString($expectedJson, $actua
...
```
Comment (sebastianbergmann) [inline src/Framework/Assert.php]: We don't use inline comments like this.

File: src/Framework/Constraint/JsonMatches.php [hunk 3]
```diff
@@ -48,19 +48,53 @@ public function __construct($value)
      */
     protected function matches($other)
     {
-        $decodedOther = json_decode($other);
-        if (json_last_error()) {
-            return false;
-        }
+		list($error, $recodedOther) = $this->_canonicalizeJson($other);
+		if ($error)
```
Comment (sebastianbergmann) [inline src/Framework/Constraint/JsonMatches.php]: The `{` must be on the same line.

File: src/Framework/Constraint/JsonMatches.php [hunk 4]
```diff
@@ -48,19 +48,53 @@ public function __construct($value)
...
+		{
+			return false;
+		}
 
-        $decodedValue = json_decode($this->value);
-        if (json_last_error()) {
-            return false;
-        }
+		list ($error, $recodedValue) = $this->_canonicalizeJson($this->value);
+		if ($error)
```
Comment (sebastianbergmann) [inline src/Framework/Constraint/JsonMatches.php]: The `{` must be on the same line.

File: src/Framework/Constraint/JsonMatches.php [hunk 5]
```diff
@@ -48,19 +48,53 @@ public function __construct($value)
...
     }
 
+	/*
+	 * To allow comparison of JSON strings, first process them into a consistent
+	 * format so that they can be compared as strings.
+	 * @return array ($error, $canonicalized_json)  The $error parameter is used
+	 * to indicate an error decoding the json.  This is used to avoid ambiguity
+	 * with JSON strings consisting entirely of 'null' or 'false'.
+	 */
+	private function _canonicalizeJson($json)
```
Comment (sebastianbergmann) [inline src/Framework/Constraint/JsonMatches.php]: We do not prefix non-public methods with `_`.

File: src/Framework/Constraint/JsonMatches.php [hunk 6]
```diff
@@ -48,19 +48,53 @@ public function __construct($value)
...
+		$reencodedJson = json_encode($decodedJson);
+		return array(false, $reencodedJson);
+	}
+
+	/*
+	 * JSON object keys are unordered while PHP array keys are ordered.
+	 * Sort all array keys to ensure both the expected and actual values have
+	 * their keys in the same order.
+	 */
+	private function _recursiveSort(&$json)
```
Comment (sebastianbergmann) [inline src/Framework/Constraint/JsonMatches.php]: We do not prefix non-public methods with `_`.

**Comment by sebastianbergmann** [comment]
Thank you for your pull request. I will merge it once you have fixed the issues I pointed out.

**Comment by kalenp** [comment]
Commits have been updated with styling feedback and an additional test case.
I have targeted this pull request to the `master` branch.  Is this correct?

**Comment by kalenp** [comment]
@sebastianbergmann Do the updated commits meet your style guidelines?

**Comment by kalenp** [comment]
@sebastianbergmann This is ready to be reviewed and merged.

**Comment by sebastianbergmann** [comment]
Dear contributor,

let me start by apologizing for not commenting and/or working on the issue you have reported or merging the pull request you have sent sooner.

PHPUnit 5.0 was released today. And today I am closing all open bug reports and pull requests for PHPUnit and its dependencies that I maintain. Please do not interpret the closing of this ticket as an insult or a lack of interest in your problem. I am sorry for any inconvenience this may cause.

If the topic of this ticket is still relevant then please open a new ticket or send a new pull request. If your ticket or pull request is about a defect then please check whether the issue still exists in PHPUnit 4.8 (which will received bug fixes until August 2016). If your ticket or pull request is about a new feature then please port your patch PHPUnit 5.0 before sending a new pull request.

I hope that today's extreme backlog grooming will allow me to respond to bug reports and pull requests in a more timely manner in the future.

Thank you for your understanding,
Sebastian

**Pull request closed without merging**